sgx/sawtooth_poet_sgx/libpoet_enclave/poet_enclave.cpp

```diff
@@ -1302,3 +1305,39 @@ size_t CalculateSealedSignupDataSize()
 {
     return sgx_calc_sealed_data_size(0, sizeof(ValidatorSignupData));
 } // CalculateSealedSignupDataSize
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+// The PoET Private Key and the monotonic counter (sequenceID) are never
+// exposed outside the enclave. This makes the resulting nonce unpredictable
+// They are both deterministic values. The conjunction of the private key
+// and the sequence number will occur only once in the lifecycle of that
+// validator registration.
+// Returns a 64 character hex formatted string
+std::string GenerateWaitCertificateNonce(
+        const uint8_t *pPoetPrivateKey,
+        const size_t keyLen,
+        const uint8_t *pSequenceID,
+        const size_t sequenceLen
+        )
+{
+    sgx_status_t ret;
+    sgx_sha256_hash_t nonce;
+    sgx_sha_state_handle_t sha_handle;
+
+    ret = sgx_sha256_init(&sha_handle);
+    sp::ThrowSgxError(ret, "Failed to generate wait certificate nonce");
+
+    ret = sgx_sha256_update(pPoetPrivateKey, keyLen, sha_handle);
+    sp::ThrowSgxError(ret, "Failed to generate wait certificate nonce");
+
+    ret = sgx_sha256_update(pSequenceID, sequenceLen, sha_handle);
+    sp::ThrowSgxError(ret, "Failed to generate wait certificate nonce");
+
+    ret = sgx_sha256_get_hash(sha_handle, &nonce);
+    sp::ThrowSgxError(ret, "Failed to generate wait certificate nonce");
+
+    ret = sgx_sha256_close(sha_handle);
+    sp::ThrowSgxError(ret, "Failed to generate wait certificate nonce");
```

Comment:
Would it be nice to have different error messages for each of these for debugging purpose?

Reply:
I agree with this suggestion.

Reply:
I chose to use one consistent message because this is one logical call `hash(psk, counter)`. Needing to check response codes on each incremental step rather than try/catch over the block is an idiosyncrasy of that API. The error will still direct debug efforts to this routine. In general I prefer less verbose logging in security sensitive modules. This follows owasp and other guidelines.

Reply:
Agree